### PR TITLE
Improvements to external storages list rendering

### PR DESCRIPTION
**Pull request by RobinMcCorkell**
This PR will introduce rendering the storages list in JS rather than in PHP, to improve performance and unify the codebase. In addition, it will display applicable global storages in the Personal config section, to a limited degree.

Ref #16307 #15575

### CODE REVIEW

**Comment by karlitschek** [comment]
looks good to me. Not fully tested. :+1:

**Comment by Xenopathic** [comment]
OK, global storages are now shown in the Personal settings, greyed out. @MTRichards does that satisfy part of the criteria in #16307 ?

cc @PVince81 @icewind1991 @MorrisJobke @jancborchardt

**Comment by Xenopathic** [comment]
It seems that mount priorities have been broken for a while... the last commit here fixes that. Also, the personal storage view will now only display the storage that is active if multiple ones exist with the same mountpoint (related to priority), e.g.:

You have two mounts with the same mountpoint, but one applies to a specific group while the other applies to all users. The specific one should always take precedence (it now does thanks to the priority fix). In the personal storages view, only the specific one will be shown, not the more general one, since only that one applies to you.

**Comment by Xenopathic** [comment]
OK, all major code is now unit tested. Get reviewing! It would be awesome to have this in 8.2, especially with the priority fix.

cc @PVince81 @icewind1991 @MorrisJobke @LukasReschke (less PHP!) @DeepDiver1975

**Comment by icewind1991** [comment]
Bug: after entering a folder name and selecting a storage type the creates the next row with the same folder name and status as the row you just picked the type for:

**Comment by Xenopathic** [comment]
Any chance of getting this merged for 8.2? cc @icewind1991 @PVince81 @DeepDiver1975

**Comment by icewind1991** [comment]
:+1:

File: apps/files_external/js/settings.js [hunk 1]
```diff
@@ -825,14 +916,14 @@ MountConfigListView.prototype = _.extend({
 	 * @return {OCA.External.StorageConfig} storage model instance
 	 */
 	getStorageConfig: function($tr) {
-		var storageId = parseInt($tr.attr('data-id'), 10);
+		var storageId = parseInt($tr.data('id'), 10);
```
Comment (PVince81) [inline apps/files_external/js/settings.js]: Why this change ? Note that `.data` auto-parses it to an integer, so makes the `parseInt` redundant.

I usually prefer to use `.attr('data-id')` because it prevents auto-parsing which can sometimes cause nasty side effects.

**Comment by PVince81** [comment]
Needs rebase due to conflicts

**Comment by PVince81** [comment]
- [ ] BUG: create SFTP storage with "RSA public key" in the personal page. After mounting it works. Then go back to the personal page, the "generate keys" button is broken:

sftp-button-broken

File: apps/files_external/js/settings.js (hunk 1 above)
Comment (Xenopathic) [inline apps/files_external/js/settings.js]: `.attr()` was used before since the PHP would spit out `data-id=""` attributes for the generated rows, but when generating everything in JS (and applying the ID with `.data('id', ...)` it often doesn't get stored as a real attribute, but in local data storage. So `attr()` doesn't work at that point.

Comment (PVince81) [inline apps/files_external/js/settings.js]: unless you apply the id as `.attr('id', value)` which has the advantage of showing up in the DOM inspector too.

Comment (Xenopathic) [inline apps/files_external/js/settings.js]: I still don't see the benefits of using `.attr()` instead of `.data()`. `.data()` is more semantically correct, and we only ever put integers into it so the auto-parsing is consistent.

**Comment by PVince81** [comment]
Oh noes, there are conflicts again :frowning:

**Comment by icewind1991** [comment]
Rebased

**Comment by PVince81** [comment]
- [ ] BUG: same as https://github.com/owncloud/core/pull/18999#issuecomment-149497966https://github.com/owncloud/core/pull/18999#issuecomment-149497966 but now when I refresh the page my mount point config does not appear at all. The mount point is visible in "All files" but not in the personal page

**Comment by icewind1991** [comment]
@Xenopathic do you have the time to finish this pr?

**Comment by Xenopathic** [comment]
@icewind1991 I'll try and finish it over the weekend... :see_no_evil:

**Comment by Xenopathic** [comment]
@PVince81 Unable to reproduce, however I've found a bug where the SSH keys will be regenerated every time the page is loaded...

**Comment by Xenopathic** [comment]
I chose to use jQuery Deferred to solve the problem. Authentication methods or backends using custom JS now get an additional parameter, a `jQuery.Deferred`, which is resolved on full construction of the config row (including setting of data). This allows the handler to directly initialise the fields it needs, for example buttons, but still defer final steps (like pre-initialising SSH keys for new storages) until everything is ready.

**Comment by PVince81** [comment]
@Xenopathic I created admin and personal mounts from master, then switched to this branch.
It doesn't show any entry in the GUI.

**Comment by PVince81** [comment]
Ok it does now, submodule fail...

**Comment by PVince81** [comment]
Looks good now :+1: 

@icewind1991 second review ?

**Comment by icewind1991** [comment]
:+1:

**Comment by PVince81** [comment]
Regression: cannot mount Dropbox https://github.com/owncloud/core/issues/20835
